File: cpp/01KnapsackDP.cpp
// knapsack_dp.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <algorithm>
#include <limits>

using namespace std;
using Clock = chrono::high_resolution_clock;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cout << "0/1 Knapsack DP\n";
    cout << "Enter number of items (n): ";
    int n;
    if (!(cin >> n) || n <= 0) {
        cerr << "Invalid number of items.\n";
        return 1;
    }

    vector<int> weight(n), value(n);
    cout << "Enter " << n << " items as: <weight> <value> (one per line)\n";
    for (int i = 0; i < n; ++i) {
        if (!(cin >> weight[i] >> value[i])) {
            cerr << "Invalid input for item " << i << ".\n";
            return 1;
        }
        if (weight[i] < 0 || value[i] < 0) {
            cerr << "Weights and values must be non-negative.\n";
            return 1;
        }
    }

    cout << "Enter knapsack capacity: ";
    int capacity;
    if (!(cin >> capacity) || capacity < 0) {
        cerr << "Invalid capacity.\n";
        return 1;
    }

    // Safety threshold for full 2D DP reconstruction:
    // (n+1) * (capacity+1) cells — each cell is an int (4 bytes). Cap near ~10M cells (~40MB).
    const long long SAFE_CELLS = 10'000'000LL;
    long long required_cells = (long long)(n + 1) * (long long)(capacity + 1);

    long long max_value = 0;
    vector<int> chosen_indices; // filled only if reconstruction performed

    // Start timing DP compute
    auto t0 = Clock::now();

    if (required_cells <= SAFE_CELLS) {
        // Full 2D DP table with reconstruction
        vector<int> table((n + 1) * (capacity + 1), 0);
        auto dp = [&](int i, int w) -> int& {
            return table[i * (capacity + 1) + w];
        };

        for (int i = 1; i <= n; ++i) {
            int wi = weight[i - 1];
            int vi = value[i - 1];
            for (int w = 0; w <= capacity; ++w) {
                int without = dp(i - 1, w);
                int with_item = (w >= wi) ? dp(i - 1, w - wi) + vi : without;
                dp(i, w) = max(without, with_item);
            }
        }

        max_value = dp(n, capacity);

        // Reconstruct chosen items
        int w = capacity;
        for (int i = n; i >= 1; --i) {
            if (dp(i, w) != dp(i - 1, w)) {
                chosen_indices.push_back(i - 1);
                w -= weight[i - 1];
                if (w <= 0) break;
            }
        }
        reverse(chosen_indices.begin(), chosen_indices.end());
    } else {
        // Memory friendly 1D DP (no reconstruction)
        vector<int> dp(capacity + 1, 0);
        for (int i = 0; i < n; ++i) {
            int wi = weight[i], vi = value[i];
            for (int c = capacity; c >= wi; --c) {
                dp[c] = max(dp[c], dp[c - wi] + vi);
            }
        }
        max_value = dp[capacity];
    }

    // End timing
    auto t1 = Clock::now();
    chrono::duration<double> elapsed = t1 - t0;

    // Output
    cout << "\nMaximum value achievable: " << max_value << "\n";
    cout << "DP execution time: " << elapsed.count() << " seconds\n";

    if (!chosen_indices.empty()) {
        cout << "\nSelected items (index : weight, value):\n";
        for (int idx : chosen_indices) {
            cout << idx << " : " << weight[idx] << " , " << value[idx] << "\n";
        }
    } else if (required_cells > SAFE_CELLS) {
        cout << "\n(Note: reconstruction skipped due to memory threshold. Reduce n or capacity to enable reconstruction.)\n";
    } else {
        cout << "\n(No items selected or reconstruction found none.)\n";
    }

    return 0;
}

File: Java/01KnapsackDP.java
// Main.java
import java.io.*;
import java.util.*;

public class Main {
    // Safety threshold for full 2D DP (cells). Adjust if you have a lot of memory.
    private static final long SAFE_CELLS = 10_000_000L; // ~40MB for int table

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        System.out.println("0/1 Knapsack DP Benchmark");
        System.out.print("Enter number of items (n): ");
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        if (n <= 0) {
            System.err.println("n must be positive.");
            return;
        }

        int[] weight = new int[n];
        int[] value  = new int[n];

        System.out.println("Enter " + n + " items as: <weight> <value> (one per line):");
        for (int i = 0; i < n; ++i) {
            st = new StringTokenizer(br.readLine());
            weight[i] = Integer.parseInt(st.nextToken());
            value[i]  = Integer.parseInt(st.nextToken());
            if (weight[i] < 0 || value[i] < 0) {
                System.err.println("Weights and values must be non-negative.");
                return;
            }
        }

        System.out.print("Enter knapsack capacity: ");
        st = new StringTokenizer(br.readLine());
        int capacity = Integer.parseInt(st.nextToken());
        if (capacity < 0) {
            System.err.println("Capacity must be non-negative.");
            return;
        }

        long requiredCells = (long)(n + 1) * (long)(capacity + 1);
        System.out.println("\n(n+1)*(capacity+1) = " + requiredCells +
            (requiredCells <= SAFE_CELLS ? " <= SAFE_CELLS -> using 2D DP with reconstruction." 
                                        : " > SAFE_CELLS -> using memory-friendly 1D DP (no reconstruction)."));

        int maxValue = 0;
        List<Integer> chosen = null; // indices chosen, if reconstructed

        // Start timing DP computation
        long t0 = System.nanoTime();

        if (requiredCells <= SAFE_CELLS) {
            // Full 2D DP using flattened int array for speed
            int cols = capacity + 1;
            int[] table = new int[(n + 1) * cols];
            // dp(i,w) -> table[i*cols + w]
            for (int i = 1; i <= n; ++i) {
                int wi = weight[i - 1];
                int vi = value[i - 1];
                int prevBase = (i - 1) * cols;
                int curBase  = i * cols;
                for (int w = 0; w <= capacity; ++w) {
                    int without = table[prevBase + w];
                    int withItem = (w >= wi) ? table[prevBase + (w - wi)] + vi : without;
                    table[curBase + w] = Math.max(without, withItem);
                }
            }
            maxValue = table[n * (capacity + 1) + capacity];

            // Reconstruct chosen items
            chosen = new ArrayList<>();
            int w = capacity;
            for (int i = n; i >= 1; --i) {
                int cur = i * (capacity + 1) + w;
                int prev = (i - 1) * (capacity + 1) + w;
                if (table[cur] != table[prev]) {
                    // item i-1 was taken
                    chosen.add(i - 1);
                    w -= weight[i - 1];
                    if (w <= 0) break;
                }
            }
            Collections.reverse(chosen);
        } else {
            // Memory-efficient 1D DP (no reconstruction)
            int[] dp = new int[capacity + 1];
            for (int i = 0; i < n; ++i) {
                int wi = weight[i];
                int vi = value[i];
                if (wi > capacity) continue;
                for (int c = capacity; c >= wi; --c) {
                    int cand = dp[c - wi] + vi;
                    if (cand > dp[c]) dp[c] = cand;
                }
            }
            maxValue = dp[capacity];
        }

        // End timing
        long t1 = System.nanoTime();
        double elapsedSeconds = (t1 - t0) / 1e9;

        // Output
        System.out.println("\nMaximum value achievable: " + maxValue);
        System.out.printf("DP execution time: %.9f seconds%n", elapsedSeconds);

        if (chosen != null) {
            if (chosen.isEmpty()) {
                System.out.println("\nNo items selected.");
            } else {
                System.out.println("\nSelected items (0-based index : weight, value):");
                int totalW = 0, totalV = 0;
                for (int idx : chosen) {
                    System.out.println(idx + " : " + weight[idx] + " , " + value[idx]);
                    totalW += weight[idx];
                    totalV += value[idx];
                }
                System.out.println("Total weight: " + totalW + ", Total value: " + totalV);
            }
        } else {
            System.out.println("\n(Decision reconstruction skipped due to memory threshold. Reduce n or capacity to enable reconstruction.)");
        }
    }
}
